src/include/sof/audio/buffer.h

```diff
@@ -221,17 +243,77 @@ static inline void buffer_lock(struct comp_buffer *buffer, uint32_t *flags)
  */
 static inline void buffer_unlock(struct comp_buffer *buffer, uint32_t flags)
 {
+	CHECK_BUFFER_CORE(buffer);
+
 	if (!buffer->inter_core)
 		return;
 
-	/* save lock pointer to avoid memory access after cache flushing */
-	spinlock_t *lock = buffer->lock;
+	/* wtb and inv to avoid buffer locking in read only situations */
+	dcache_writeback_invalidate_region(buffer, sizeof(*buffer));
+
+	spin_unlock_irq(&buffer->lock, flags);
+}
+
+#if CONFIG_CAVS && CONFIG_CORE_COUNT > 1
+__must_check static inline struct comp_buffer *buffer_acquire(
+	struct comp_buffer *buffer, uint32_t *flags)
+{
+	/* assert is someone passes a cached address in here. */
+	assert(!is_uncached(buffer));
+
+	CHECK_BUFFER_CORE(buffer);
+
+	if (!buffer->inter_core) {
+		/* Ignored by buffer_unlock() below, silences "may be
+		 * used uninitialized" warning.
+		 */
+		*flags = 0xffffffff;
+		goto out;
+	}
+
+	/* Expands to: *flags = ... */
+	spin_lock_irq(&buffer->lock, *flags);
+
+	/* invalidate in case something has changed during our wait */
+	dcache_invalidate_region(uncache_to_cache(buffer), sizeof(*buffer));
+
+out:
+	/* client can now use cached buffer */
+	return uncache_to_cache(buffer);
+}
+
+__must_check static inline struct comp_buffer *buffer_release(
```

Comment:
does it really need to return a pointer again? You aren't supposed to use the object after releasing it, are you? In follow up patches in this series the return value is never used.

Reply:
no, this was going to be removed anyway. We only have limited use cases where it would help.

Reply:
@lgirdwood @lyakh I have some changes with the fixes suggested here. I will take over the PR from here.